docs/operators/deduper.md

```diff
@@ -0,0 +1,523 @@
+|:------:|:-----------------------------:|:----------:|:----------:|:----------------:|
+|1|10|10|1|Not Expired|
+|2|20|20|11|Not Expired|
+|3|25|25|16|Not Expired|
+|4|40|40|31|Not Expired|
+|5|21|40|31|Expired|
+|6|35|40|31|Not Expired|
+|7|45|45|36|Not Expired|
+|8|57|57|48|Not Expired|
+
+
+### Time Buckets (A component of Managed State)
+
+One of the requirements of the Deduper is to store all the unique
+tuples (actually, just the keys of tuples). Keeping an ever growing
+cache in memory is not scalable. So what we need is a limited cache
+backed by a persistent store. Now to reduce cache misses, we load a
+chunk of data (called Buckets), together into memory. Buckets help
+narrow down the search of duplicates for incoming tuples. A Bucket is an
+abstraction for a collection of tuples all of which share a common hash
+value based on some hash function or a range of time, for example: a
+bucket of data for5 contiguous minutes. A Bucket has a span property called Bucket Span.
+
+### Bucket Span
+
+Bucket span is simply the range of the domain
+that is covered by the Bucket. This span is specified in
+the domain of the Expiry key. If the Expiry
+Key is time,  then the Bucket span
+will be specified in seconds. It is
+only defined in case tuples have an Expiry Key.
+
+### Number of Buckets
+
+The number of buckets can be given by - Num Buckets = Expiry
+Period / Bucket Span.
+
+This is because at any point of time, we need only store Expiry
+Period worth of data.
+
+### Example - Buckets
+
+![](images/deduper/image01.png)
+
+Assumptions
+------------------------
+
+### Assumption 1 <a name="Assumption1"></a>
+
+This assumption is only applicable in case of Dedup with
+Expiry.
+
+For any two tuples, t1 and t2 having dedup keys d1 and d2, and
+expiry keys e1 and e2, respectively, the following holds:
+
+If d1 = d2, then e1 = e2
+
+In other words, there may never
+be two tuples t1 and t2 such that:
+
+Tuple 1: d1, e1
+
+Tuple 2: d2, e2
+
+d1 = d2 and e1 != e2
+
+This assumption was made with respect to certain use cases. These
+use cases follow this assumption in that the records which are
+duplicates are exactly identical. An example use case is when log
+messages are replayed erroneously, and we want to identify the duplicate
+log messages. In such cases, we need not worry about two different log
+messages having the same identifier but different timestamps. Since its
+a replay of the same data, the duplicate records are assumed to be
+exactly identical.
+
+Flow of a Tuple through Dedup Operator
+---------------------------------------------------
+
+Tuples flow through the Dedup operator one by one. Deduper may
+choose to process a tuple immediately, or store it in some data
+structure for later processing. We break down the processing of the
+Deduper by various stages as follows.
+
+### Deduper View
+
+A tuple always arrives at the input
+port of the Dedup operator. Once arrived, the Deduper does
+the following tasks.
```

Comment:
Combine both sentences: When a tuple arrives at the input port of the Dedup operator, it does￼ the following tasks.